### PR TITLE
[Filebeat] Add option for S3 input to work without SQS notification

### PR DESCRIPTION
**Pull request by aspacca**
## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Comment by aspacca** [comment]
fixes #26934 18205

File: x-pack/filebeat/input/awss3/input.go [hunk 1]
```diff
@@ -8,6 +8,8 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/elastic/beats/v7/libbeat/monitoring"
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/input.go:11]: this can go around line 20 below.

File: x-pack/filebeat/input/awss3/input.go [hunk 2]
```diff
@@ -62,13 +63,28 @@ func (in *s3Input) Test(ctx v2.TestContext) error {
...
+	var err error
+	var collector s3Collector
+
+	metricRegistry := monitoring.GetNamespace("dataset").GetRegistry()
+	inputMetrics := newInputMetrics(metricRegistry, ctx.ID)
+
+	if in.config.QueueURL != "" {
+		collector, err = in.createSQSCollector(ctx, pipeline, inputMetrics)
+		if err != nil {
+			return err
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/input.go:75]: maybe we should add some extra info wrap around this err to know which collector the error is from.

File: x-pack/filebeat/input/awss3/s3_integration_test.go [hunk 3]
```diff
@@ -75,7 +77,44 @@ func getConfigForTest(t *testing.T) config {
 	return config
 }
 
-func defaultTestConfig() *common.Config {
+// getConfigForTestBucketCollector function gets aws credentials for integration tests.
+func getConfigForTestBucketCollector(t *testing.T) config {
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/s3_integration_test.go:81]: This is probably better to be relocated into a common location so other inputs like aws-cloudwatch can also use for testing. Similar to what we have in Metricbeat: https://github.com/elastic/beats/blob/master/x-pack/metricbeat/module/aws/mtest/integration.go#L14

File: x-pack/filebeat/input/awss3/input.go (hunk 2 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/input.go:75]: 👍

File: x-pack/filebeat/input/awss3/s3_integration_test.go (hunk 3 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/s3_integration_test.go:81]: the only problem is https://github.com/elastic/beats/blob/master/x-pack/filebeat/input/awss3/config.go#L20
`config` is private and won't be accessible from an eventual `x-pack/filebeat/input/ftest/integration.go`
What do you suggest? I'm not sure making it public is fine

**Comment by aspacca** [comment]
>I have made corresponding change to the default configuration files

@kaiyan-sheng do you think we should add `s3_bucket` variable for aws modules?

File: CHANGELOG.next.asciidoc [hunk 4]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (kaiyan-sheng) [inline CHANGELOG.next.asciidoc:1]: extra line here?

File: x-pack/filebeat/input/awss3/collector.go [hunk 5]
```diff
@@ -21,12 +21,15 @@ import (
 	"sync"
 	"time"
 
+	"github.com/elastic/beats/v7/libbeat/statestore"
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/collector.go:24]: this should be moved down around line 34

File: x-pack/filebeat/input/awss3/input.go [hunk 6]
```diff
@@ -7,48 +7,128 @@ package awss3
 import (
 	"context"
 	"fmt"
+	"time"
+
+	"github.com/urso/sderr"
+
+	"github.com/elastic/beats/v7/libbeat/statestore"
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/input.go:14]: also here

File: x-pack/filebeat/input/awss3/collector.go [hunk 7]
```diff
@@ -93,7 +114,110 @@ type s3Context struct {
...
+
+	if err := paginator.Err(); err != nil {
+		return nil, errors.Wrap(err, "error FilterLogEvents with Paginator")
+	}
+
+	return s3Infos, nil
+}
+
+func (c *s3BucketCollector) run() {
+	defer c.logger.Info("s3 input worker has stopped.")
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/collector.go:177]: should we also defer here in the info log between s3bucket collector vs s3sqs collector (line 221)?

Comment (aspacca) [reply x-pack/filebeat/input/awss3/collector.go:177]:
```
2021-07-27T10:28:48.376+0200    INFO    [input.aws-s3]  awss3/collector.go:201  s3 input worker has stopped.    {"id": "4F2455F48AA183A4", "s3_bucket": "arn:aws:s3:::mybucket"}
```
vs
```
2021-07-27T10:28:48.376+0200    INFO    [input.aws-s3]  awss3/collector.go:242  s3 input worker has stopped.    {"id": "4F2455F48AA183A4", "queue_url": "https://sqs.myregion.amazonaws.com/123456/myqueue"}

```

but I can add extra information about the collector type in the message

File: CHANGELOG.next.asciidoc (hunk 4 above)
Comment (aspacca) [reply CHANGELOG.next.asciidoc:1]: indeed

File: x-pack/filebeat/input/awss3/collector.go (hunk 5 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/collector.go:24]: I have a file watcher on goland that should take care of ordering imports, that sometimes fails
but I would expect mage update would take care of this as well

File: x-pack/filebeat/module/aws/_meta/docs.asciidoc [hunk 8]
```diff
@@ -128,6 +146,24 @@ Default to be 300 seconds.
...
+AWS S3 bucket ARN (Required when `var.queue_url` is not set).
+
+*`var.s3_bucket_poll_interval`*::
+
+Interval between list requests to the S3 bucket. Default to be 120 seconds.
+
+*`var.s3_bucket_objects_expiration`*::
+
+Expiration of the S3 objects in the bucket. If you have an expiration set on the
+S3 objects in the bucket you can set it here so that internal status of the object will
```
Comment (exekias) [inline x-pack/filebeat/module/aws/_meta/docs.asciidoc:160]: I didn't know this exists, nice. Do you have any docs on the AWS side we can reference? Also, do you know how common it is to use this as compared to just "keep everything" in the bucket? I guess this has some implications on our thinking around the state store.

I'm mostly concerned about forever growing state in high traffic scenarios

Comment (aspacca) [reply x-pack/filebeat/module/aws/_meta/docs.asciidoc:160]: https://aws.amazon.com/it/blogs/aws/amazon-s3-object-expiration/

I will add the link to docs as well

I'm not sure about how much is common indeed. The problem with not expiring objects, once they are ingested, if we set a TTL in order to flush from the state store they would be ingested again.
It is a trade off between repeated ingestions and having a state store that keeps growing (I wrote in the documentation that this has impact on memory usage of filebeat).
If ingestion of the same data is idempotent, we could define this TTL as "forget status" for the s3 object, explain better the implications and how the best setup is to link it to the s3 object expiration.
But what if the data were purged from elastic? After a little time they would reappear and this can be confusing for the users

File: x-pack/filebeat/input/awss3/input.go [hunk 9]
```diff
@@ -62,13 +95,54 @@ func (in *s3Input) Test(ctx v2.TestContext) error {
 }
 
 func (in *s3Input) Run(ctx v2.Context, pipeline beat.Pipeline) error {
-	collector, err := in.createCollector(ctx, pipeline)
+	var err error
+	var collector s3Collector
+
+	persistentStore, err := in.store.Access()
+	if err != nil {
+		return sderr.Wrap(err, "Can not access persistent store")
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/input.go:103]: Is it required to use `"github.com/urso/sderr"` here? Same as Line 109.

File: x-pack/filebeat/input/awss3/clean_store.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	"github.com/elastic/beats/v7/libbeat/statestore"
+	"github.com/elastic/go-concert/timed"
+	"github.com/elastic/go-concert/unison"
+)
+
+func cleanStore(canceler unison.Canceler, logger *logp.Logger, store *statestore.Store, states *States, interval time.Duration, objectExpiration time.Duration) {
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/awss3/clean_store.go:17]: nit: maybe add some comment for this function as well?

File: x-pack/filebeat/input/awss3/input.go (hunk 9 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/input.go]: I didn't realise I introduced it, moved to `fmt.Errorf` like the rest

File: x-pack/filebeat/input/awss3/clean_store.go (hunk 10 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/clean_store.go:18]: done

**Review by kaiyan-sheng — APPROVED** [review]

File: x-pack/filebeat/module/aws/_meta/docs.asciidoc (hunk 8 above)
Comment (exekias) [reply x-pack/filebeat/module/aws/_meta/docs.asciidoc:166]: how about doing keeping a time based cursor after each run? Once a full object list / retrieval is completed we could potentially remove all the states and replace them with a single timestamp mark meaning: anything older than this has been shipped already

Comment (aspacca) [reply x-pack/filebeat/module/aws/_meta/docs.asciidoc:166]: State is checked on listing: https://github.com/elastic/beats/blob/db7e44cfd921cc8f853f18d60305d8df6ae3365d/x-pack/filebeat/input/awss3/collector.go#L140-L142

If I got it right, we should add here that even if it is not new (in this case because the state was purge because stored) but `state.LastModified` is before the last listing we should skip anyway because it is shipped already.
We should have a specific check on state to ensure that the previous state we compare doesn't exists.
Yes, I think it would be a great improvement.

Comment (exekias) [reply x-pack/filebeat/module/aws/_meta/docs.asciidoc:166]: > we should add here that even if it is not new

I think you actually mean "new". but yes, I believe we are in agreement.

To me this approach raises the question: should he have the TTL option at all? or just rely on doing cleanup + keeping this `lastModified` cursor after each complete fetch?

File: x-pack/filebeat/input/awss3/clean_store.go [hunk 11]
```diff
@@ -38,13 +42,17 @@ func gcStore(logger *logp.Logger, started time.Time, store *statestore.Store, st
 	if err := gcClean(store, states, keys); err != nil {
 		logger.Errorf("Failed to remove all entries from the registry: %+v", err)
 	}
+
+	if err := store.Set(awsS3WriteCommitStateKey, commitWriteState{time.Now()}); err != nil {
```
Comment (exekias) [inline x-pack/filebeat/input/awss3/clean_store.go:46]: this should not be `time.Now()` (dependant on the machine running Filebeat), it should be extracted from the listed objects (`lastModified` field), keeping the newest date. Trusting the local machine to be in sync to S3 dates can lead to sync problems, where the local time may be wrong, or timezones differ.

Comment (aspacca) [reply x-pack/filebeat/input/awss3/clean_store.go:46]: good catch!

File: x-pack/filebeat/input/awss3/input.go [hunk 12]
```diff
@@ -62,13 +94,55 @@ func (in *s3Input) Test(ctx v2.TestContext) error {
...
+		}
+	}
+
+	err = in.grp.Go(func(canceler unison.Canceler) error {
+		interval := in.store.CleanupInterval()
+		if interval <= 0 {
+			interval = 5 * time.Minute
+		}
+
+		interval = 1 * time.Minute
```
Comment (exekias) [inline x-pack/filebeat/input/awss3/input.go:134]: this last line invalidates all the previous

File: x-pack/filebeat/input/awss3/input.go [hunk 13]
```diff
@@ -62,13 +94,55 @@ func (in *s3Input) Test(ctx v2.TestContext) error {
...
+		cleanStore(canceler, ctx.Logger, persistentStore, states, interval)
```
Comment (exekias) [inline x-pack/filebeat/input/awss3/input.go:135]: Should this be run for the SQS Collector? I think that one doesn't use the state store at all

Comment (exekias) [inline x-pack/filebeat/input/awss3/input.go:135]: I've been thinking about this cleanup logic, can we really separate bucket listing from cleanup? If we use `lastModified` dates for cleanup, somehow we need to guarantee that cleanup happens **immediately after** a full listing, otherwise we are at risk of skipping not-yet-listed objects that have a `lastModified` time older than what we saw so far.

File: x-pack/filebeat/input/awss3/input.go (hunk 12 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/input.go:134]: leftover from debugging

File: x-pack/filebeat/input/awss3/input.go (hunk 13 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/input.go:135]: I've moved in the if where we instantiate the S3Bucket collector

Comment (aspacca) [reply x-pack/filebeat/input/awss3/input.go:135]: I see the problem, but I think then that to solve this we should bind cleanup not to bucket listing but to the mark-as-stored step.
The problem I see is that mark-as-stored is done at object level, and not for the whole listing: and still we will end up in the same situation where we skip not-yet-listed objects that have a `lastModified` older than what we marked in the mark-as-stored step.

We are trying to infer a state (or the outcome of a missing state: skip the object) from a state that's missing in store.

Another option is to keep states by listing operation, and remove the whole batch at once when all of them are processed/marked-as-stored. In this case we have a precise batch we know that either we have a state for we can compare to, or we miss and we should not skip. What do you think?

Comment (exekias) [reply x-pack/filebeat/input/awss3/input.go:135]: > Another option is to keep states by listing operation, and remove the whole batch at once when all of them are processed/marked-as-stored. In this case we have a precise batch we know that either we have a state for we can compare to, or we miss and we should not skip. What do you think?

This sounds reasonable to me, a few concerns:
- A later list operation should also be able to invalidate older states (right?)
- Would memory usage be an issue? I think it wouldn't but may be missing something

Comment (aspacca) [reply x-pack/filebeat/input/awss3/input.go:135]: >A later list operation should also be able to invalidate older states (right?)

I was thinking about keeping "buckets" of list operations, once all the state in a bucket are marked as stored the bucket will be purged. When we check if an object can be skipped we check in every buckets of the list operations.
We could reduce to O(n) (n = number of list operations) with the right algorithm

> Would memory usage be an issue? I think it wouldn't but may be missing something

Memory wise we will increase the requirements by a factor of `s3_bucket_poll_interval/cleanup_interval`. Since the default values are respectively 120s and 300s we will something between 2x and 3x

Comment (exekias) [reply x-pack/filebeat/input/awss3/input.go:135]: > I was thinking about keeping "buckets" of list operations, once all the state in a bucket are marked as stored the bucket will be purged. When we check if an object can be skipped we check in every buckets of the list operations.

Can 2 listings happen at the same time? They are run in loop, keeping the state for all of them presents some issues, for instance, interrupted listings (for instance if FB dies). I'm wondering if it would be just fine to care about the last successful listing. Always doing a cleanup after a it, removing previous states.

File: x-pack/filebeat/input/awss3/state.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+func NewState(bucket, key string, size int64, lastModified time.Time) State {
+	s := State{
+		Bucket:       bucket,
+		Key:          key,
+		LastModified: lastModified,
+		Size:         size,
+		Stored:       false,
+	}
+
+	s.Id = s.Bucket + s.Key + strconv.FormatInt(s.Size, 10) + s.LastModified.String()
```
Comment (andrewkroh) [inline x-pack/filebeat/input/awss3/state.go:33]: Have you considered using the ETag rather than size and last modified? The ETag should only change when the content changes which could make it useful as part of a unique ID.

File: x-pack/metricbeat/modules.d/gcp.yml.disabled [hunk 15]
```diff
@@ -44,6 +44,20 @@
...
+  credentials_file_path: "your JSON credentials file path"
+  exclude_labels: false
+  period: 1m
+  metrics:
+    - aligner: ALIGN_NONE
+      service: kubernetes
+      service_metric_prefix: kubernetes.io/
+      metric_types:
+        - "container/cpu/core_usage_time"
+
```
Comment (exekias) [inline x-pack/metricbeat/modules.d/gcp.yml.disabled:60]: this seems unrelated to this pR

File: x-pack/osquerybeat/Jenkinsfile.yml [hunk 16]
```diff
@@ -19,8 +19,11 @@ stages:
           make -C x-pack/osquerybeat update;
           make check-no-changes;
         stage: lint
-    build:
-        mage: "mage build test"
+    unitTest:
+        mage: "mage build unitTest"
+        stage: mandatory
+    goIntegTest:
+        mage: "mage goIntegTest"
```
Comment (exekias) [inline x-pack/osquerybeat/Jenkinsfile.yml:26]: unrelated' I see a couple of commits around testing that don't belong here I think

File: x-pack/filebeat/input/awss3/state.go (hunk 14 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/state.go:33]: much better, yes, I will change to `Etag` for equalness, `LastMofidied` we still need for states tracking

File: x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc [hunk 17]
```diff
@@ -231,6 +231,7 @@ Specific AWS permissions are required for IAM user to access SQS and S3:
 
 ----
 s3:GetObject
+s3:ListBucket
```
Comment (andrewkroh) [inline x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc:234]: I think we should separate the permissions required by usage mode (whether using it with sqs or just s3 alone).

File: filebeat/docs/modules/aws.asciidoc [hunk 18]
```diff
@@ -15,8 +15,8 @@ This file is generated! See scripts/docs_collector.py
 beta[]
 
 This is a module for aws logs. It uses filebeat s3 input to get log files from
-AWS S3 buckets with SQS notification. This module supports reading s3 server
-access logs with `s3access` fileset, ELB access logs with `elb` fileset, VPC
+AWS S3 buckets with SQS notification or directly polling list of S3 objects in an S3 bucket.
+This module supports reading s3 server access logs with `s3access` fileset, ELB access logs with `elb` fileset, VPC
 flow logs with `vpcflow` fileset, and CloudTrail logs with `cloudtrail` fileset.
```
Comment (andrewkroh) [inline filebeat/docs/modules/aws.asciidoc:21]: I think the docs should explain trade-offs and use cases for the two "modes". Like steer users towards SQS S3 notifications primarily and then suggest S3 listing for back-filling existing logs.

File: x-pack/filebeat/input/awss3/states.go [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+}
+
+type listingInfo struct {
+	totObjects    int
+	storedObjects int
+}
+
+// States handles list of s3 object state. One must use NewStates to instantiate a
+// file states registry. Using the zero-value is not safe.
+type States struct {
```
Comment (andrewkroh) [inline x-pack/filebeat/input/awss3/states.go:63]: Does this need to be exported from the package? If there are no external users then I'd make it unexported. Same for `State`.

File: x-pack/filebeat/input/awss3/state.go [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+// State is used to communicate the reading state of a file
+type State struct {
+	Id           string    `json:"id" struct:"id"`
+	Bucket       string    `json:"bucket" struct:"bucket"`
+	Key          string    `json:"key" struct:"key"`
+	LastModified time.Time `json:"last_modified" struct:"last_modifed"`
+	Size         int64     `json:"size" struct:"size"`
+	Stored       bool      `json:"stored" struct:"stored"`
```
Comment (andrewkroh) [inline x-pack/filebeat/input/awss3/state.go:20]: Please add a godoc comment about what "Stored" indicates. I think it means that the object's contents have been published and ACKed by the output.

File: x-pack/filebeat/input/awss3/state.go [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+)
+
+// State is used to communicate the reading state of a file
+type State struct {
+	Id           string    `json:"id" struct:"id"`
```
Comment (andrewkroh) [inline x-pack/filebeat/input/awss3/state.go:15]:
```suggestion
	ID           string    `json:"id" struct:"id"`
```

**Review by andrewkroh — COMMENTED** [review]
This will definitely be a useful enhancement. I know I have a bunch of old S3 access logs that I want to ingest.

I think several of my comments won't be relevant after we merge our two PRs and if so ignore them.

File: x-pack/filebeat/input/awss3/collector.go [hunk 22]
```diff
@@ -93,7 +130,260 @@ type s3Context struct {
...
+			defer wg.Done()
+			c.getS3Objects(s3InfoChan, errChan)
+		}(wg, s3InfoChan, errChan)
+
+		for i := 0; i < c.config.S3BucketNumberOfWorkers; i++ {
+			wg.Add(1)
+			go func(wg *sync.WaitGroup, s3InfoChan <-chan *s3Info, errChan <-chan error) {
+				defer wg.Done()
+				c.process(s3InfoChan, errChan)
+				c.purge()
```
[the remark this answers is not in the document]
Comment (aspacca) [reply x-pack/filebeat/input/awss3/collector.go:360]: `purge` is concurrency safe, the lock is acquired in `getS3Objects` (https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-0717dc8e6158cdc32ee79b54fe8fe67ea2ebad820d5935016e4f9a18d5cdd153R162) and released either in the same method (if no objects were collected: https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-0717dc8e6158cdc32ee79b54fe8fe67ea2ebad820d5935016e4f9a18d5cdd153R228) or when all the objects in the listing are stored (https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-34a1e64be4068f5099b1af6dc062ac9afd7f3f229041e25caeba0efa7f475e20R52)

Probably `purge` could be called after all the worker goroutine exits, since we will have the state in memory and the same object from different listing won't be processed twice anyway, but if filebeat dies or is restarted when some of the goroutine exited and some not yet we will not have the states synched to disk so the objects would be processed again when filebeat is restarted.

I was thinking about how much performance impact has synching the store to disk so often. Once moved the logic on your PR I will write a benchmark and see.

File: x-pack/filebeat/input/awss3/collector.go [hunk 23]
```diff
@@ -93,7 +130,260 @@ type s3Context struct {
...
+			errC := make(chan error, 1)
+			err := handleS3Objects(c.s3, []s3Info{*info}, errC, c.cancellation, c.config.APITimeout, c.publisher, c.metrics, c.logger)
+			if err != nil {
+				c.logger.Error(fmt.Errorf("handleS3Objects failed: %w", err))
+				break processingLoop
+			}
+		case err := <-errChan:
+			if err != nil {
+				c.logger.Error("S3 get objects failed: ", err)
+				break processingLoop
```
[the remark this answers is not in the document]
Comment (aspacca) [reply x-pack/filebeat/input/awss3/collector.go:267]: good catch: before introducing https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-0717dc8e6158cdc32ee79b54fe8fe67ea2ebad820d5935016e4f9a18d5cdd153R154 we could have only one error per `getS3Objects` call, but the method could still collect some objects to process, so the `process` goroutines that don't receive the error should keep running.
actually there's no need to shutdown any of them when they receive an error but just log it (being able to receive multiple errors they should not be stopped indeed).
they will be shutdown when there's no more message in the channel here https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-0717dc8e6158cdc32ee79b54fe8fe67ea2ebad820d5935016e4f9a18d5cdd153R245.
what do you think?

File: x-pack/filebeat/input/awss3/collector.go [hunk 24]
```diff
@@ -93,7 +130,260 @@ type s3Context struct {
...
+			}
+
+			state := NewState(info.name, info.key, info.size, info.lastModified)
+			//check if another worker already is on it
+			dummyValue := struct{}{}
+			_, loaded := c.workersProcessingMap.LoadOrStore(state.Id, dummyValue)
+			if loaded {
+				continue
+			} else {
+				defer c.workersProcessingMap.Delete(state.Id)
```
[the remark this answers is not in the document]
Comment (aspacca) [reply x-pack/filebeat/input/awss3/collector.go:255]: good catch, I want to clean up after each iteration

File: x-pack/filebeat/input/awss3/collector.go [hunk 25]
```diff
@@ -93,7 +130,260 @@ type s3Context struct {
...
+		}
+
+		// lock for the listing page and state in workersListingMap
+		// this map is shared with the storedOp and will be unlocked there
+		lock.Lock()
+		c.workersListingMap.Store(listingID.String(), lock)
+
+		s3Infos := make([]*s3Info, 0)
+		page := paginator.CurrentPage()
+		for _, object := range page.Contents {
```
[the remark this answers is not in the document]
Comment (aspacca) [reply x-pack/filebeat/input/awss3/collector.go:167]: I was thinking to add metrics for it as well in the refactored code.

File: x-pack/filebeat/input/awss3/collector.go [hunk 26]
```diff
@@ -409,10 +714,27 @@ func (c *s3Collector) createEventsFromS3Info(svc s3iface.ClientAPI, info s3Info,
 	// Decode JSON documents when content-type is "application/json" or expand_event_list_from_field is given in config
 	if resp.ContentType != nil && *resp.ContentType == "application/json" || info.ExpandEventListFromField != "" {
 		decoder := json.NewDecoder(bodyReader)
-		err := c.decodeJSON(decoder, objectHash, info, s3Ctx)
+		logEvents, err := decodeJSON(decoder, info, logger)
```
[the remark this answers is not in the document]
Comment (aspacca) [reply x-pack/filebeat/input/awss3/collector.go:717]: >Basically you'll know if the file was fully ingested if ProcessS3Object returns a nil error and the eventACKTracker.Wait()

`eventACKTracker` will be shared by all objects in the listing, `eventACKTracker.Wait()` returning will tell me if every object in the listing was fully read, not a single one. Am I missing something?

Now I receive some metadata related to the object (including its source listing), and then I have a states handler in the ackHandler (https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-589195ecff44ed2c6bf097da676822765ae86c2e73fc5789aaf0a64d8bb26ff1R242) reacting to full read of every object.

If I want to keep the same model the `eventACKTracker` should receive the metadata as well (here: https://github.com/elastic/beats/pull/27199/files#diff-c6357a5d68a965123a3b3dc3cbf0bddfe7a5628dceb5327885da33f1735a3facR302?), and In the ackHandler I should be able to know the if listing was full read with a non blocking call (ie: not `eventACKTracker.Wait()`)

On the refactored code I was thinking about having an `s3Lister` like there is an `sqsReceiver`, an `s3ListingProcessor` like an `sqsProcessor` and an `s3ListingProcessor.processS3Object` like an `sqsProcessor.processS3Events`. 

How do you see if I move extract states handling from the ackHandler and have it in `s3ListingProcessor.processS3Object`? In the end here I will process an object a time, and not multiple ones like for `sqsProcessor.processS3Events`

File: x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc (hunk 17 above)
Comment (aspacca) [reply x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc:234]: fine for me

File: filebeat/docs/modules/aws.asciidoc (hunk 18 above)
Comment (aspacca) [reply filebeat/docs/modules/aws.asciidoc:21]: Oki. Actually there are cases where the SQS S3 notifications could not be setup at all, that was the main reason for introducing a direct S3 bucket polling.

File: x-pack/filebeat/input/awss3/states.go (hunk 19 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/states.go:63]: 👍

File: x-pack/filebeat/input/awss3/state.go (hunk 20 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/state.go:20]: 👍

File: x-pack/filebeat/input/awss3/state.go (hunk 21 above)
Comment (aspacca) [reply x-pack/filebeat/input/awss3/state.go:15]: there is a mage target to apply gofmt, is there?

File: x-pack/filebeat/input/awss3/collector.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (andrewkroh) [reply x-pack/filebeat/input/awss3/collector.go:717]: > eventACKTracker will be shared by all objects in the listing

You can create an `eventACKTracker` per S3 object. This way you can know when that object's data was fully ACKed.

BTW I'm going to add more godoc to the interface methods in my PR. But I'd expect you'd use it like this.

```
s3ObjectHandler := s3ObjectHandlerFactory.Create(s3ObjectCoordinates)
if s3ObjectHandler == nil {
  // None of the configured file selectors matched this file so just skip it.
  return / continue
}


acker := newEventACKTracker(ctx)
if err := s3ObjectHandler.ProcessS3Object(); err != nil {
  // Not sure how you want to handle this case. This means that it failed to process some aspect of the object.
  // Could be that the GetObject failed completely or that the JSON was malformed, for example.
  log.Warn(...)
}

// All events that were generated from reading the object are now ACKed by the output.
acker.Wait()

// Update the state tracking for this one object.
```
